### PR TITLE
[v1alpha2 Migration] add v1alpha2 controllers

### PR DESCRIPTION
**Pull request by sbueringer**
Initial migration to v1alpha2 controllers as described here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/380#issuecomment-522018403

(This also contains the commits from: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pulls)

### CODE REVIEW

**Comment by sbueringer** [comment]
Dev env for v1alpha2 is described here: https://github.com/sbueringer/capi-dev

**Comment by sbueringer** [comment]
Seems to work :):

````
Every 1.0s: openstack server list                                                                                                                                                                                                localhost.localdomain: Sun Aug 18 20:06:43 2019

+--------------------------------------+-------------------------+--------+---------------------------------------------------+-----------------+--------+
| ID                                   | Name                    | Status | Networks                                          | Image           | Flavor |
+--------------------------------------+-------------------------+--------+---------------------------------------------------+-----------------+--------+
| 77e3ae08-e8be-4342-98d7-927ff074733b | ix1pi020-kube-node-01   | ACTIVE | k8s-clusterapi-cluster-default-ix1pi020=10.6.0.6  | coreos-2023.5.0 | large  |
| 7c983a77-2640-41b4-af77-792a864295e2 | ix1pi020-kube-master-02 | ACTIVE | k8s-clusterapi-cluster-default-ix1pi020=10.6.0.3  | coreos-2023.5.0 | large  |
| 55ae4749-4862-4548-a302-fc182f31f45f | ix1pi020-kube-master-01 | ACTIVE | k8s-clusterapi-cluster-default-ix1pi020=10.6.0.17 | coreos-2023.5.0 | large  |
+--------------------------------------+-------------------------+--------+---------------------------------------------------+-----------------+--------+


Every 1.0s: kubectl --kubeconfig /home/fedora/.kube/kind-config-kind --context kubernetes-admin@kind --namespace default get machine,openstackmachine                                                                            localhost.localdomain: Sun Aug 18 20:06:30 2019

NAME                                               PROVIDERID                                           PHASE
machine.cluster.x-k8s.io/ix1pi020-kube-master-01   openstack:////55ae4749-4862-4548-a302-fc182f31f45f   provisioned
machine.cluster.x-k8s.io/ix1pi020-kube-master-02   openstack:////7c983a77-2640-41b4-af77-792a864295e2   provisioned
machine.cluster.x-k8s.io/ix1pi020-kube-node-01     openstack:////77e3ae08-e8be-4342-98d7-927ff074733b   provisioned

NAME                                                                       AGE
openstackmachine.infrastructure.cluster.x-k8s.io/ix1pi020-kube-master-01   23m
openstackmachine.infrastructure.cluster.x-k8s.io/ix1pi020-kube-master-02   6m44s
openstackmachine.infrastructure.cluster.x-k8s.io/ix1pi020-kube-node-01     6m44s

NAME                      STATUS     ROLES    AGE   VERSION
ix1pi020-kube-master-01   NotReady   master   20m   v1.15.0-1+702333862a992f-dhc
ix1pi020-kube-master-02   NotReady   master   92s   v1.15.0-1+702333862a992f-dhc
ix1pi020-kube-node-01     NotReady   node     27s   v1.15.0-1+702333862a992f-dhc

````

**Comment by sbueringer** [comment]
Implement is mostly done (HA cluster works in my case). Documentation to use it is here: https://github.com/sbueringer/capi-dev

**Comment by sbueringer** [comment]
/assign @hidekazuna 
/assign @jichenjc 

Can you please take a look? :)

**Comment by sbueringer** [comment]
Overall this seems to work in my env. Code should also be okay. I'll have some followup task like cleaning up the Makefile & refactor the logging a bit. Depending on how seen this PR is reviewed I would integrate it here or create an additional PR for that

**Comment by jichenjc** [comment]
I will try on ubuntu env as well, so might need your help and confirm
1) 445 already merged into this PR? 
2) do I need Tilt ? I am not using IDE and no experience on it, is it mandatory for non-IDE env? 
3) I saw a set of samples changed , I assume that's the thing I need migrate from my existing one to (the format and name etc) ,correct?

**Comment by sbueringer** [comment]
> I will try on ubuntu env as well, so might need your help and confirm
> 
> 1. 445 already merged into this PR?
> 2. do I need Tilt ? I am not using IDE and no experience on it, is it mandatory for non-IDE env?
> 3. I saw a set of samples changed , I assume that's the thing I need migrate from my existing one to (the format and name etc) ,correct?

1. It was the basis for this PR. So yes it's included
2. I also just know it since 3 days :). Documentation what it does is below
3. Yes the new samples from the samples folder are the files you need in addition to the stuff Tilt deploys

Tilt basically does the following:
* Docker Images: build, load into Kind & live-update the following images:
  * cluster api controller
  * cluster api bootstrap provider kubeadm (not needed yet)
  * cluster api Openstack controller
* It seds the names of loaded Images in the `config/default` folder of each of these 3 repos (cluster-api, cluster-api-bootstrap-provider-kubeadm, cluster-api-provider-openstack)
* it executes kustomize on `/config/default` and pipes the result to kubectl apply
* It redeploys as soon as something is changed

Problem is that clusterctl is not migrated to v1alpha2 and there isn't much documentation yet. Most of the documentation probably will be done for clusterctl or in the common ClusterAPI I guess.

**Comment by jichenjc** [comment]
ok, struggled for a while and now the tilt is up including all containers (a set of minor changes...)
now I will deploy from ubuntu and I will share some comments/info after the test

File: samples/templates/clouds-secrets/configs/clouds.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+clouds:
+  os1pi020:
+    auth:
+      username: E415_s_os1pi020
+      password: KJFs!3CuCu%kjYuyfd87
```
Comment (jichenjc) [inline samples/templates/clouds-secrets/configs/clouds.yaml]: hope this is not your real password @sbueringer   :)

**Comment by jichenjc** [comment]
I can't comment on cluster-api-provider-openstack/samples/templates/clouds-secrets/kustomize.yaml (don't know the reason), but I made some modification to make namespace to capo-system, otherwise the container can't find the secret during startup, you find same issue? 

---
apiVersion: v1
data:
  cacert:  ........
  clouds.yaml: ......
kind: Secret
metadata:
  name: cloud-config
+  namespace: capo-system
type: Opaque

  namespace: capo-system

File: samples/templates/clouds-secrets/configs/clouds.yaml (hunk 1 above)
Comment (sbueringer) [reply samples/templates/clouds-secrets/configs/clouds.yaml]: Thought I removed it but I will change it 😀

**Comment by sbueringer** [comment]
> I can't comment on cluster-api-provider-openstack/samples/templates/clouds-secrets/kustomize.yaml (don't know the reason), but I made some modification to make namespace to capo-system, otherwise the container can't find the secret during startup, you find same issue?
> 
> ```
> apiVersion: v1
> data:
>   cacert:  ........
>   clouds.yaml: ......
> kind: Secret
> metadata:
>   name: cloud-config
> +  namespace: capo-system
> type: Opaque
> ```

Oh yes you're right. I fixed it in the PR. Problem was I was using a copy of the samples (with completely different user-data) and I only fixed it there

**Comment by jichenjc** [comment]
@sbueringer seems KUBELET_VERSION={{ .Machine.Spec.Versions.Kubelet }} in master/worker node need to be updated? there is no Versions anymore

and the namespace settings is fairly complicated now... makes test very hard :(

**Comment by sbueringer** [comment]
> @sbueringer seems KUBELET_VERSION={{ .Machine.Spec.Versions.Kubelet }} in master/worker node need to be updated? there is no Versions anymore
> 
> and the namespace settings is fairly complicated now... makes test very hard :(

I'll take a look

What do you mean with namespace settings?

**Comment by jichenjc** [comment]
the machine spec, cluster spec, secret ...all need set namespace now ... they don't need in v1alpha1, nothing special, I am still fixing those namespace settings  on my system

**Comment by sbueringer** [comment]
> the machine spec, cluster spec, secret ...all need set namespace now ... they don't need in v1alpha1, nothing special, I am still fixing those namespace settings on my system

Oh yes. I think it will be easier in the future. clusterctl should be working soon again and there are more tools planned. Not sure what the exact scope of clusteradm will be

**Comment by sbueringer** [comment]
> > @sbueringer seems KUBELET_VERSION={{ .Machine.Spec.Versions.Kubelet }} in master/worker node need to be updated? there is no Versions anymore

Yup the changed Versions.Kubelet to just Version. I fixed it. (I need Ubuntu to test stuff like this :/)

**Comment by jichenjc** [comment]
ok, I made some modifications locally and ubuntu works now. so I think it's time to merge this and evolve from here :), I will submit PR to fix ubuntu cases.

root@jj-master:~/capi-dev/cluster-api-provider-openstack/samples/output# kubectl get machine
NAME                             PROVIDERID                                           PHASE
test-cluster-kube-master-01      openstack:////d6b5c2d1-4b0c-4196-a8fc-d8f6d973bdd0   provisioned
test-cluster-openstack-node-01   openstack:////c6ef9117-3f9b-4e46-904f-283ca5d174fa   provisioned


cloudusr@test-cluster-kube-master-01:~$ kubectl get nodes
NAME                             STATUS   ROLES    AGE   VERSION
test-cluster-kube-master-01      Ready    master   85m   v1.15.0
test-cluster-openstack-node-01   Ready    node     40m   v1.15.0

**Pull request merged**